xdelta3-dir-patcher

```diff
@@ -21,6 +49,355 @@ class AttributeDict(dict):
+
+        for future in concurrent.futures.as_completed(self.futures):
+            if future.exception():
+                raise future.exception()
+
+        # Leftover runners might have again clobbered the output
+        self._fix_terminal()
+
+        # XXX: If nothing was ran, we might not have a start time
+        if not self.start_time:
+            self.start_time = time.time()
+
+        print('Runner time: %.2fs' % (time.time() - self.start_time))
+
+# ---------------------------- ARCHIVE ADAPTERS ----------------------------
+class XDeltaArchive(object):
+    def __init__(self, archive_path):
+        self.archive_object = XDeltaArchive.get_archive_instance(archive_path)
+
+    def __enter__(self):
+        return self.archive_object
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.archive_object.close()
+
+    @staticmethod
+    def get_archive_instance(archive_path):
+        for clazz in XDelta3AbstractArchiveImpl.__subclasses__():
+            if clazz.can_open(archive_path):
+                return clazz(archive_path)
+
+        raise RuntimeError('Error! Archive %s bad or not supported!' % archive_path)
+
+class XDelta3AbstractArchiveImpl(object):
+    def __init__(self):
+        self.lock = threading.RLock()
+
+    def _acquire_lock(self):
+        self.lock.acquire()
+
+    def _release_lock(self):
+        self.lock.release()
+
+class XDelta3FsImpl(XDelta3AbstractArchiveImpl):
+    def __init__(self, path, for_writing = False):
+        super().__init__()
+
+        self.path = path
+        self.items = None
+
+    # Placeholders to allow us to use 'with` keywords on this implementation
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        pass
+
+    def close(self):
+        pass
+
+    @staticmethod
+    def can_open(archive):
+        return path.isdir(archive)
+
+    def list_files(self):
+        if self.items:
+            return self.items
+
+        print('FS: Gathering filelist')
+        items = []
+        for root, dirs, filenames in walk(self.path):
+            for directory in dirs:
+                absolute_path = path.join(root, directory)
+
+                # Strip prefix so the files look relative
+                relative_path = path.relpath(absolute_path, self.path) + path.sep
+
+                items.append(relative_path)
+
+            for filename in filenames:
+                absolute_path = path.join(root, filename)
+
+                # Strip prefix so the files look relative
+                relative_path = path.relpath(absolute_path, self.path)
+
+                items.append(relative_path)
+
+        print('FS: Gathering completed')
+
+        self.items = items
+
+        return self.items
+
+    def expand(self, filename, extraction_path):
+        target_path = path.join(extraction_path, filename)
+        source_path = path.join(self.path, filename)
+
+        if not target_path.endswith(path.sep):
```

Comment:
@dbnicholson
Re: First comment
- The idea is that we key off the destination path and not source. If we did it from `source_path`, we might get into strange situations where there's both a file and folder with the same name and the algo just wouldn't know which one of those you meant to expand.

Re: Second comment
- Big problem there is that some implementations are devoid of notions about some items that we return from list_files (e.g. Tarfile might not have any folder info at all but just files with internal paths) but to be fair those can be worked around. The pure strings approach is easy to work with though, hides the underlying implementation details, and requires a smaller memory footprint. If I'm not mistaken, both `tarfile` and `zipfile` also cache the internal members so we would only reduce the internal lib hash lookup and the concurrency problems are on tarfile archive file pointer so it wouldn't really fix that either.